src/segment/segment_reader.rs

```diff
@@ -376,6 +381,30 @@ impl<R: Read + Seek> SegmentReader<R> {
                     self.created,
                     id
                 );
+                let dur = docstore.get_duration().unwrap_or(0 as i32);
+                if dur < self.duration {
+                    //Refresh cache
+                    debug!(
+                        "Docstore cache entry will be refreshed as it has old duration {}, new: {} {:?} for shard: {} segment: {} block: {}",
+                        dur,
+                        self.duration,
+                        SystemTime::now().duration_since(curr_time).unwrap(),
+                        self.shard_id,
+                        self.created,
+                        id
+                    );
+                    let new_d = self.get_docstore(id)?;
```

Comment:
We have to release the lock before calling `self.get_docstore()` and acquire it again.